### ISSUE TITLE
Support boot on ukvm

### ISSUE THREAD

**Issue by nyh** [cloudius-systems/osv#886]
Would be nice to support booting OSv on ukvm, the unikernel hypervisor (https://github.com/Solo5/solo5/tree/master/ukvm) which doesn't emulate a normal PC with PCI, serial port, 16-bit boot - etc., but rather something simpler and faster booter. For an extra challenge, can we build one image which can boot on both (probably the boot sector will need to somehow figure out what we're running under)?

See http://blog.includeos.org/2017/06/23/popek-goldberg-machines-considered-harmful for a nice example of what could be done with ukvm

**Comment by wkozaczuk** [comment]
Could you please elaborate what more specifically had to be done to make OSv bootable on uKVM?What new device drivers do we need to add?

**Comment by nyh** [comment]
Yes, it seems like a non-trivial exercise, I don't know if it's worth the effort. It seems (as far as I searched) that not even Linux has drivers for UKVM.
The UKVM tree includes an example guest kernel ("Solo5") which has network, disk, console, etc., drivers - see https://github.com/Solo5/solo5/tree/master/kernel/ukvm (it has a BSD-like license so we could copy code from it). It's probably even deeper than that, UKVM probably also changes PCI, parts of the bios we assume, the boot process, etc., so a lot of different parts of OSv may need to change to support it.

**Comment by wkozaczuk** [comment]
For those interested I opened related uKVM issue - https://github.com/Solo5/solo5/issues/249. There is a lot of relevant information. I may not appreciate all the info there but the fact uKVM does not provide interrupts and preemptiveness makes me think that making OSv run on uKVM may be even more difficult or even impossible.

**Comment by foreverpersist** [comment]
I think it is difficult to run OSv on ukvm.  How about using ukvm's design idea to accelerate OSv?
Is it possible to customize qemu to support a simple but efficient IO back-end like ukvm and append a specific front-end on OSv for it?
